core: Provide all C11 atomic operations

gebart opened the pull request:
This PR adds the remaining `__atomic` library functions for C11 support and also provides the `__sync` functions for older GCC versions (msp430).

All atomic.h uses in the tree have been replaced by their C11 equivalents.

The functions provided here are only used when the compiler doesn't know how to do a lockless atomic operation for the given target architecture. Therefore, with a recent GCC, most of these functions are only used on Cortex-M0, MSP430.

**Note:** This PR works fine with `-std=gnu99` (the current setting), even though the title says C11

gebart commented:
- fixes #4700
 - replaces #4888

Kijewski commented on core/atomic_c11.c:
```diff
@@ -0,0 +1,363 @@
+    (void) memorder;
+    unsigned int mask = irq_disable();
+    memcpy(ret, ptr, size);
+    memcpy(ptr, val, size);
```
No line `memcpy(val, ret, size);`?

gebart replied:
It is implemented as the manual specifies it:

https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html

gebart commented:
msp430-libc was missing stdatomic.h so I copied Newlib's header and the required support headers to msp430-common (BSD license).

Full C11 atomic functionality is still provided when compiling with `-std=gnu99`, at least for the newlib platforms, so we don't need to switch to `-std=gnu11` yet.

kaspar030 commented:
Nice! Did you do some codesize comparisons?

gebart commented:
@kaspar030 I did not, it's difficult to find anything to compare since the unit tests had to be modified. The emb6 test app saved 28 bytes of ROM on samr21-xpro. Almost nothing was using atomic.h anymore before this PR so its removal was well deserved.

DipSwitch commented:
Does this override the GCC buildins which for ARM use LDREX and STREX by default?

gebart commented:
@DipSwitch nope, GCC only generates library calls if there is no lock-free hardware implementation available.

You can verify by building the tests-core in `tests/unittests` and inspecting the disassembly (`make tests-core && arm-none-eabi-objdump -d bin/$BOARD/tests-core/tests-core-atomic.o`)

josephnoir commented:
Admittedly, I don't understand most of the changes and thus have a hard time reviewing the changes.

gebart commented:
@josephnoir that's OK, I'll reassign.
@kaspar030 do you mind taking a look at this?

miri64 commented:
Kaspar is on vacation, so I need to postpone this one :(

gebart commented:
Ping @kaspar030 
Any other maintainer who would be willing to review this in time for christmas?
It cleans up the implementation quite a lot and it is not overly complex, despite the seemingly large change set.

gebart commented:
ping maintainers!
Can we get this merged in time for 2017.01?

OlegHahm commented:
@gebart, I will try to help.

OlegHahm commented on core/atomic_c11.c:
```diff
@@ -0,0 +1,363 @@
+ *
+ * @param[in]  n         width of the data, in bytes
+ */
+#define TEMPLATE_ATOMIC_LOAD_N(n) \
```
I agree that macros make sense here, but just for the record: I hate macros and if they will ever create any problems - I told you so! ;-)

gebart replied:
So do I :smiley:
In this case, like you said, it makes sense to reduce the amount of copy+pasta in this file. Expanding the templates in source would only make it more likely that any future bug fix would miss some instance and make the implemented functions diverge.

gebart commented:
ping @OlegHahm @kaspar030 
Murdock is happier now. Only needs squashing.
The Doxygen warning in core/mutex.h is unrelated.

OlegHahm commented on sys/posix/pthread/pthread_spin.c:
```diff
@@ -37,7 +39,7 @@ int pthread_spin_destroy(pthread_spinlock_t *lock)
     }
 
-    (void) lock;
+    atomic_flag_clear(&(lock->flag));
```
This is basically a semantic change (from doing nothing to doing something), right? So, I would propose to have a separate commit for this.

gebart replied:
Will split into a separate commit.

gebart replied:
After reviewing the manual pages again (http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_spin_destroy.html), I see that this change was not needed and should just be removed.

gebart commented:
maybe we could copy the clz/ctz implementations from libgcc to a separate file until msp430 is updated to a modern version?

OlegHahm commented:
Sounds reasonable.

gebart commented:
I did import the libgcc function at first, but it was licensed under GPL-3 and I don't want to open that can, so I added my own naive implementation of clz and ctz. Untested code, compiles fine though.
@OlegHahm could you run the unittests/tests-core on an MSP430 board please?

gebart commented:
rebased on top of #6395

gebart commented:
Rebased, removed the unrelated change in pthread_spin.c.
I think this should be ready for merging as soon as it has been tested on at least one board from the different families (msp430, avr, cortex-m0)

gebart commented:
@OlegHahm do you mind taking a look at this PR again? I'd like to merge this now before any other PRs cause rebase conflicts, since this touches several different parts of the system.

gebart commented:
Needs to run the core unittests on an msp430 board.

gebart commented:
@OlegHahm ping!

kaspar030 commented:
I just tested the core unittests on msb-430h, looks good!

```
2017-02-08 16:37:06,321 - INFO # RIOT MSP430 hardware initialization complete.
2017-02-08 16:37:06,322 - INFO #
2017-02-08 16:37:06,330 - INFO # main(): This is RIOT! (Version: 2017.04-devel-131-gf0551-6b9ed7c519ae-pr5688)
2017-02-08 16:37:08,213 - INFO # .....................................................................................
2017-02-08 16:37:08,216 - INFO # OK (85 tests)
```

gebart commented:
Fixes #6491 
(to let GitHub auto-close the issue)

gebart commented:
Core change needs a second ACK. @melshuber, @OlegHahm, @Kijewski? 

@kaspar030 Thanks, I have felt responsible for this since I was the one who introduced the old incomplete atomic.h API. It feels nice to finally make the support complete, and also standards compliant.

The resulting assembly code for a simple load is quite heavy (see https://github.com/RIOT-OS/RIOT/issues/6491#issuecomment-275701000), but I guess that's an optimization we can postpone until it becomes an issue.

gebart commented:
@melshuber I would prefer to not leave the old implementation in, since it was not really working as intended (`ATOMIC_VALUE` on avr for example)

melshuber commented:
@gebart well enough, I just wanted to mention it
